playground/CustomResources/CustomResources.AppHost/TestResource.cs

```diff
@@ -0,0 +1,73 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.Dashboard;
+using Aspire.Hosting.Lifecycle;
+
+static class TestResourceExtensions
+{
+    public static IResourceBuilder<TestResource> AddTestResource(this IDistributedApplicationBuilder builder, string name)
+    {
+        builder.Services.AddLifecycleHook<TestResourceLifecycleHook>();
+
+        var rb = builder.AddResource(new TestResource(name))
+                      .WithDashboardState(state => state with
+                      {
+                          ResourceType = "Test Resource",
+                          State = "Starting",
+                          Properties = [
+                              ("P1", "P2"),
+                              (DashboardKnownProperties.Source, "Custom")
+                          ]
+                      })
+                      .ExcludeFromManifest();
+
+        return rb;
+    }
+}
+
+internal sealed class TestResourceLifecycleHook : IDistributedApplicationLifecycleHook, IAsyncDisposable
+{
+    private readonly CancellationTokenSource _tokenSource = new();
+
+    public Task BeforeStartAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken = default)
+    {
+        foreach (var item in appModel.Resources.OfType<TestResource>())
+        {
+            if (item.TryGetLastAnnotation<DashboardAnnotation>(out var annotation))
+            {
+                var states = new[] { "Starting", "Running", "Finished" };
```

Comment:
This sample brings up some interesting questions. Should it be possible to wait for the dashboard state of a resource to be initialized? I think yes

Reply:
I'm liking this model overall. I think we should be more generic in our terminology here, I can see this annotation turning into a bit of a proxy for not only state, but also commanding against specific resources.